docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,126 @@
+
+### Deleting Profiles
+
+A profile can only be deleted from metadata when the profile is not associated to any other EdgeX object to include device, provision watcher, event or reading.
+
+### Association Changes
+
+Today, a device profile is associated to a device via its name (the ProfileName on the Device struct).  The profile can also associated to events and readings (via the ProfileName in the Event and Reading objects).
+
+Profiles can be associated to more than one device.  Profiles can be associated to multiple events/readings.
+
+Rules about change of a device profile association have never been explicitly defined.  While the section above outline what can/cannot be changed in the profile and under what circumstances, this section outlines whether a device, event or reading can alter the associated device profile and under what circumstances.
+
+Change of the device profile on a device – never allowed unless there are no events or readings in the system.  There is too much risk that an existing event or reading also contains the device profile reference – causing complexity in knowing what information is in the event/reading.
+
+A provision watcher also has an association to a device profile.  However, the device profile is only used to know the type of device to create in case of provisioning.  Therefore, if a device profile changes, it does not impact the provision watcher.
+
+A device profile cannot be deleted while it is associated to a provision watcher
+
+Change of the device profile on an event or reading is never allowed.  Events and readings are immutable.
+
+### Applicability
+
+This document dictates the rules around allowed change to device profiles (and associations).  All EdgeX services will abide by these rules (implementing appropriate checks and validation as necessary to enforce the rules) when this ADR is approved to include:
+- service APIs
+- user interfaces (like the Edge GUI)
+- client tools (like the EdgeX CLI)
+
+## Proposed Design
+
+The following metadata API changes are suggested as a means to implement the device profile rules of this ADR
+
+- Block Device DELETE when there is any associated event/reading exists.
+- Block Profile (Upload) PUT when there is any associated device/event/reading exists.
+- Block Profile DELETE when there is any associated device/event/reading exists.
+- Allow empty Profile POST (containing no device resources or commands)
+- Add Profile General Property PATCH API
+- Add Profile Device Resource POST API
+- Add Profile Device Resource PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Resource DELETE API (check it's not associated to any device, reading or event)
+- Add Profile Device Command POST API
+- Add Profile Device Command PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Command DELETE API (check it's not associated to any device, reading or event)
```

Comment:
If we already have a PUT API for profiles, why do we need to add all these individual APIs? Can't someone do all of the above actions via profile PUT?

Reply:
Or Patch??

Reply:
The whole PUT API will be blocked if there is any Device associated to the Profile.
Those small APIs could be applied to Device Profile which is associated to the existing Devices.

Reply:
Could we not separate validation logic from the API?  Meaning that we have one API PUT, and if someone attempts to change elements of a profile that are not allowed to be changed, then the PUT is rejected.  But if someone called the API to just set the Description, for example, it would be allowed.  The API remains constant, but the validation logic underneath determines whether the PUT is accepted and sends appropriate error message when it it is rejected.

Reply:
If we use only one PUT API, users have to send the whole profile every time for modifying one device resource description.
If we add a single profile PATCH API, Device Resource and Device Command are sub-object, according to the convention of our PATCH API, users should still send all the Device Resources to update one description.